New Control: Tabview

tznind opened the pull request:
Hey.  I have created a new control that replicates tabs (like browsers, notepad++ etc).

I have added 2 scenarios

 - TabViewExample.cs demonstrates a range of tabs (100 tabs)
 - Notepad.cs demonstrates creating/removing tabs dynamically

![tabs](https://user-images.githubusercontent.com/31306100/111185344-c5d57500-85a9-11eb-854c-3c5f74b15bd8.gif)

tig commented:
AWESOME!!!

BDisp commented:
@tznind to avoid the `ValidatePosDim` error and force changes on `Pos` or `Dim`, it's necessary to set the Layout to `Absolute` first. See the `Update` method in the Button.cs. I made this to avoid break the `Layout.Computed` inadvertently.

tznind commented:
> @tznind to avoid the `ValidatePosDim` error and force changes on `Pos` or `Dim`, it's necessary to set the Layout to `Absolute` first. See the `Update` method in the Button.cs. I made this to avoid break the `Layout.Computed` inadvertently.

Cool, I thought it was something like that.

I can change it to 
```
tabsBar.LayoutStyle = LayoutStyle.Absolute;
tabsBar.Y = 0;
tabsBar.LayoutStyle = LayoutStyle.Computed;
```

or I can leave it how it is (and remove comment about `ValidatePosDim`)?
```
tabs.Y = Pos.Percent(0);
```

Let me know which you would prefer or if there is a better approach.

BDisp commented:
You can maintain the comment about `ValidatePosDim` to alert others. I should alert in the exception to change from `Computed` to `Absolute` if it is the case. May be later if it another change is needed.
You can do in the way you like more. In the `Button.cs` I save the Layout to a variable and restore it again. If you think the `Percent` works well, leave it as it is. You are doing a great job. Thanks.

tznind commented:
Cool, thanks for explaining, it makes sense.  I will leave it as `Percent.Pos` then.

TabView has 2 private subviews.  One for the line of the tabs (text and tab outline) and one for the content area (inside the current tab frame).

The `ApplyStyleChanges` method (which is where `Pos.Percent` code is) all relates to these private views so shouldn't run into any state based issues imposed by the API user (i.e. the layout of the TabView itself should not matter).

tig commented on UICatalog/Scenarios/TabViewExample.cs:
```diff
@@ -0,0 +1,210 @@
+					new MenuItem ("_Quit", "", () => Quit()),
+				}),
+				new MenuBarItem ("_View", new MenuItem [] {
+					miShowTopLine = new MenuItem ("_Show Top Line", "", () => ShowTopLine()){
```
Note how you use different words to describe this flag in UI ("Show Top Line") vs. code (`ShowHeaderOverline). Please consider how this would all be simpler if you used consistent terminology. I've always found that "what makes sense to the end-user will make sense to the programmer" (not the other way around).

tig commented:
I pulled it down locally and played. Lovely!!!!

Some Feedback:

* The Notepad sample when loaded by default looks confusing. There's just a horizontal line across the screen and it looks like a bug/painting error. This is because you are not following the "New" codepath at  startup.  
  * Related to this, I think a default 'empty' Tab, with no name should still LOOK like a tab. Right?
  * I really find that when "Show Top Line" is off (which Notepad uses by default) looks confusing. I understand how it 'wastes' a row, but I'd be surprised if many apps actually use it. Consider just removing this style initially in the spirit of removing  complexity from "Minimally Loveable Product" (MLP) release.


More in review feedback.

tig commented on UICatalog/Scenarios/TabViewExample.cs:
```diff
@@ -0,0 +1,210 @@
+
+
+			frameRight.Add (new TextView () {
+				Text = @"This demos the tabs control
```
When I run this I see 

```
This demos the tabs contro
Switch between tabs using cursor key
```
(the `l` in `control` is not drawn)

tig commented on Terminal.Gui/Views/TabView.cs:
```diff
@@ -0,0 +1,794 @@
+				}
+			}
+
+			public override bool MouseEvent (MouseEvent me)
```
I can't get the left/right arrows to scroll tabs in the `Tab View` sample with mouse (in cmd.exe, which should work).

tig commented on UICatalog/Scenarios/TabViewExample.cs:
```diff
@@ -0,0 +1,210 @@
+
+
+			frameBelow.Add (new TextView () {
+				Text = @"This frame exists to check you can still tab here
```
Same here (`her`) is displayed insetead of (`here`).

tznind replied:
Think this was down to using `@""` notation with newlines in source.  I fixed by changing to regular string `"some\ntext"`

tig commented on Terminal.Gui/Views/TabView.cs:
```diff
@@ -0,0 +1,794 @@
+
+		/// <summary>
+		/// True to show the top lip of tabs.  False to directly begin with tab text during 
+		/// rendering.  When true header line occupies 3 pixels, when false only 2.
```
We use the terms `rows` and `cols` or `columns`, not pixels. Please be consistent.

tig commented on Terminal.Gui/Views/TabView.cs:
```diff
@@ -0,0 +1,794 @@
+		/// Defaults to true.
+		/// 
+		/// <para>When <see cref="TabsOnBottom"/> is enabled this instead applies to the
+		///  bottomost line of the control</para>
```
bottommost (two m)

tig commented on Terminal.Gui/Views/TabView.cs:
```diff
@@ -0,0 +1,794 @@
+
+		/// <summary>
+		/// The maximum number of characters to render in a Tab header.  This prevents one long tab 
+		/// from pushing out all the others. Values less than 0 are ignored
```
If values less than 0 are ignored, why not make the type a `uint`?

tig commented on Terminal.Gui/Views/TabView.cs:
```diff
@@ -0,0 +1,794 @@
+		/// </summary>
+		/// <param name="tab"></param>
+		/// <param name="dispose">True to dispose of the tabs control</param>
+		public void RemoveTab (Tab tab, bool dispose)
```
I wonder (I have no idea if my intution is correct, so please debate) if it would just be simpler for you to have the documentation say "caller is responsible for disposing any Tabs added". 

You don't normally see APIs on .NET classes that include a `dispose` option.

tig commented on UICatalog/Scenarios/Notepad.cs:
```diff
@@ -0,0 +1,242 @@
+			var menu = new MenuBar (new MenuBarItem [] {
+				new MenuBarItem ("_File", new MenuItem [] {
+					new MenuItem ("_New", "", () => New()),
+					new MenuItem ("_Open", "", () => Open()),
```
When I choose this two bad things happen:

1. The `Ok` button in the fIle open dialog doesn't say `Ok`
2. If I ht `tab` 5 times the app crashes.

I think you're doing something wrong in your File Open dialog calling code. I can't reproduce this using the `Editor` scenario.

tig replied:
Note "Save As" seems to work fine.

tznind replied:
Think it was calling the blank constructor that was causing this problem.  I resolved it by using the strings constructor.

BDisp replied:
@tznind please check if it will also throws exception in the `UICatalog` `All Views Tester`.

tznind replied:
Ok have figured out what I did wrong.  I used `FileDialog` instead of `OpenDialog`.  But it does seem that there is a bug when using `FileDialog()` blank constructor that results in the Open button having no text and crashing when Tabbed to.  Do you want me to raise a seperate issue for that?

tig replied:
Yes, I think you should.

tig reviewed (changes requested):
Please add unit tests.

I've added some additional review comments in the code. most of them are just suggestions for your consideration. 

This is really amazing work and I'm excited to be able to merge it in!!!! Thank you!!!!

tznind commented:
Thanks for the indepth review! I think I have fixed all the issues you raised.  Let me know if I missed any or misinterpreted any.

**When there are no Tabs**
Now when there are no Tabs the TabView its behaviour is:

- If `ShowBorder` is on then the border is rendered but the rest is blank
- If `ShowBorder` is off then control is completely blank

I thought about adding a blank tab in the case when there are no tabs but thought that would be confusing and not consistent with the other views like ListView which is just blank when there are no items.

I have added a call to `New()` at the start of the Notepad Scenario so it looks better when it first appears (consistent with notepad++ etc)

**ShowTopLine**

I have removed the setting of `ShowTopLine=false`  from the `Notepad` scenario.  

The default value is true for `ShowTopLine`. 

I think there are still some corner cases where it could be useful to cut down on the amount of vertical space taken up by TabView (e.g. when you want to show 2 tab views one above the other).  Let me know if you really don't want it in and I can delete the code for it.

tig commented:
I chatted with @migueldeicaza and we both agree this should be merged! Well done!
